akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -426,104 +424,273 @@ object StreamLayout {
-    else tryOnComplete(subscriptionStatus.get.asInstanceOf[Subscriber[T]])
+  override def onError(t: Throwable): Unit = {
+    @tailrec def rec(ex: Throwable): Unit =
+      get() match {
+        case null =>
+          if (!compareAndSet(null, ErrorPublisher(ex, "failed-VirtualProcessor"))) rec(ex)
+          else if (t == null) throw ex
+        case s: Subscription =>
+          if (!compareAndSet(s, ErrorPublisher(ex, "failed-VirtualProcessor"))) rec(ex)
+          else if (t == null) throw ex
+        case Both(s) =>
+          set(Inert)
+          try tryOnError(s, ex)
+          finally if (t == null) throw ex // must throw NPE, rule 2:13
+        case s: Subscriber[_] => // spec violation
+          getAndSet(Inert) match {
+            case Inert => // nothing to be done
+            case _     => ErrorPublisher(ex, "failed-VirtualProcessor").subscribe(s)
+          }
+        case _ => // spec violation or cancellation race, but nothing we can do
+      }
 
-  override def onNext(t: T): Unit = {
-    requireNonNullElement(t)
-    tryOnNext(subscriptionStatus.get.asInstanceOf[Subscriber[T]], t)
+    val ex = if (t == null) exceptionMustNotBeNullException else t
+    rec(ex)
   }
 
-  private final class Sub(s: Subscription) extends AtomicLong with Subscription {
-    override def cancel(): Unit = {
-      subscriptionStatus.set(InertSubscriber)
-      s.cancel()
+  @tailrec override final def onComplete(): Unit =
+    get() match {
+      case null            => if (!compareAndSet(null, EmptyPublisher)) onComplete()
+      case s: Subscription => if (!compareAndSet(s, EmptyPublisher)) onComplete()
+      case Both(s) =>
+        set(Inert)
+        tryOnComplete(s)
+      case s: Subscriber[_] => // spec violation
+        set(Inert)
+        EmptyPublisher.subscribe(s)
+      case _ => // spec violation or cancellation race, but nothing we can do
+    }
+
+  override def onNext(t: T): Unit =
+    if (t == null) {
+      val ex = elementMustNotBeNullException
+      @tailrec def rec(): Unit =
+        get() match {
+          case x @ (null | _: Subscription) => if (!compareAndSet(x, ErrorPublisher(ex, "failed-VirtualProcessor"))) rec()
+          case s: Subscriber[_]             => try s.onError(ex) catch { case NonFatal(_) => } finally set(Inert)
+          case Both(s)                      => try s.onError(ex) catch { case NonFatal(_) => } finally set(Inert)
+          case _                            => // spec violation or cancellation race, but nothing we can do
+        }
+      rec()
+      throw ex // must throw NPE, rule 2:13
+    } else {
+      @tailrec def rec(): Unit =
+        get() match {
+          case Both(s) =>
+            try s.onNext(t)
+            catch {
+              case NonFatal(e) =>
+                set(Inert)
+                throw new IllegalStateException("Subscriber threw exception, this is in violation of rule 2:13", e)
```

Comment:
Yes, but we cannot use them to signal failures because their main purpose is to avoid passing them into onError—which reminds me of a possibly fatal flaw: if an ActorGraphInterpreter shuts down because of a spec violation then this precise mechanism may lead to non-errored (i.e. leaked) Subscribers. Opened #19978.